modules/omx/omx.c

```diff
@@ -0,0 +1,300 @@
+/**
+ * @file omx.c     Raspberry Pi VideoCoreIV OpenMAX interface
+ *
+ * Copyright (C) 2016 - 2017 Creytiv.com
+ * Copyright (C) 2016 - 2017 Jonathan Sieber
+ */
+
+#include "omx.h"
+
+#include <re/re.h>
+#include <rem/rem.h>
+#include <baresip.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+#include <pthread.h>
+
+/* Avoids a VideoCore header warning about clock_gettime() */
+#include <time.h>
+#include <sys/time.h>
+
+/**
+ * @defgroup omx
+ *
+ * TODO:
+ *  * Proper sync OMX events across threads, instead of busy waiting
+ */
+
+static const int VIDEO_RENDER_PORT = 90;
+
+static int EventHandler(OMX_HANDLETYPE hComponent, void* pAppData,
+	OMX_EVENTTYPE eEvent, OMX_U32 nData1, OMX_U32 nData2,
+	void* pEventData)
+{
+	(void) hComponent;
+	switch (eEvent) {
+		case OMX_EventCmdComplete:
+			printf("omx.EventHandler: Previous command completed\n"
+				"d1=%x\td2=%x\teventData=%p\tappdata=%p\n",
+				nData1, nData2, pEventData, pAppData);
+			/* TODO: Put these event into a multithreaded queue,
+			 * properly wait for them in the issuing code */
+			break;
+		case OMX_EventError:
+			warning("omx.EventHandler: Error event type "
+				"data1=%x\tdata2=%x\n", nData1, nData2);
+			break;
+		default:
+			warning("omx.EventHandler: Unknown event type %d\t"
+				"data1=%x data2=%x\n", eEvent, nData1, nData2);
+			return -1;
+			break;
+	}
+	return 0;
+}
+
+static int EmptyBufferDone(OMX_HANDLETYPE hComponent, void* pAppData,
+	OMX_BUFFERHEADERTYPE* pBuffer)
+{
+	(void) hComponent;
+	(void) pAppData;
+	(void) pBuffer;
+
+	/* TODO: Wrap every call that can generate an event,
+	 * and panic if an unexpected event arrives */
+	return 0;
+}
+
+static OMX_ERRORTYPE FillBufferDone(OMX_HANDLETYPE hComponent,
+	OMX_PTR pAppData, OMX_BUFFERHEADERTYPE* pBuffer)
+{
+	(void) hComponent;
+	(void) pAppData;
+	(void) pBuffer;
+	puts("FillBufferDone2");
+	return 0;
+}
+
+static struct OMX_CALLBACKTYPE callbacks = {
+	&EventHandler,
+	&EmptyBufferDone,
+	&FillBufferDone
+};
+
+int omx_init(struct omx_state* st)
+{
+	OMX_ERRORTYPE err;
+
+	bcm_host_init();
+
+	st->buffers = NULL;
+
+   	pthread_mutex_init(&st->omx_mutex, 0);
+	pthread_mutex_lock(&st->omx_mutex);
+
+	err = OMX_Init();
+	err |= OMX_GetHandle(&st->video_render,
+		"OMX.broadcom.video_render", 0, &callbacks);
+
+	if (err != OMX_ERROR_NONE) {
+		error("Failed to create OMX video_render component");
+	}
+	else {
+		info("created video_render component");
+	}
+
+	pthread_mutex_unlock(&st->omx_mutex);
```

Comment:
You are right. I'm not even sure why the locking code is there. Probably it's just a relict from the previous attempt to do the OMX calls in a separate thread (which might be necessary for implementation of codec or camera capture).
I guess it's safe to assume that baresip's viddisp callbacks don't overlap/multithread?
Later today i will try to remove the locks entirely and test it.